engine/lib/generic_printer/generic_printer.ml

```diff
@@ -435,37 +435,44 @@ module Make (F : Features.T) = struct
           ~is_struct ~fields ~base =
-        match constructor with
-        | `Concrete constructor ->
-            let constructor =
-              self#_do_not_override_lazy_of_concrete_ident
-                AstPos_expr'_Construct_constructor constructor
-            in
-            let fields =
-              List.map
-                ~f:(fun field ->
-                  let name, expr = field#v in
-                  let name =
-                    match name with
-                    | `Concrete name -> name
-                    | _ ->
-                        self#assertion_failure
-                          "expr'.Construct: field: non-`Concrete"
-                  in
-                  ( self#_do_not_override_lazy_of_concrete_ident
-                      AstPos_expr'_Construct_fields name,
-                    expr ))
-                fields
-            in
-            self#expr'_Construct_inductive ~super ~constructor ~is_record
-              ~is_struct ~fields ~base
-        | `TupleCons _ ->
-            let components = List.map ~f:(fun field -> snd field#v) fields in
-            self#expr'_Construct_tuple ~super ~components
-        | `Primitive _ | `TupleType _ | `TupleField _ | `Projector _ ->
-            self#assertion_failure "Construct unexpected constructors"
+        let fields_or_empty add_space =
+          if List.is_empty fields then empty
+          else
+            add_space
+            ^^ parens
+                 (separate_map (comma ^^ space) (fun x -> (snd x#v)#p) fields)
+        in
+        match (constructor, fields, base) with
+        | `TupleCons 0, [], _ -> string "tt"
+        | `TupleCons 1, [ e ], _ ->
+            let _, e' = e#v in
+            string "(* tuple_cons 1 *)" ^^ e'#p
+        | `TupleCons _, _, None -> fields_or_empty empty
+        | _, _, _ ->
+            if is_record && is_struct then
+              match base with
+              | Some x -> string "Build_" ^^ x#p ^^ fields_or_empty space
+              | None ->
+                  string "Build_t_"
+                  ^^ (self#_do_not_override_lazy_of_global_ident
+                        Generated_generic_printer_base
+                        .AstPos_pat'_PConstruct_constructor constructor)
+                       #p ^^ fields_or_empty space
+            else if not is_record then
+              if is_struct then
+                string "Build_t_"
+                ^^ (self#_do_not_override_lazy_of_global_ident
+                      Generated_generic_printer_base
+                      .AstPos_pat'_PConstruct_constructor constructor)
+                     #p ^^ fields_or_empty space
+              else
+                (self#_do_not_override_lazy_of_global_ident
+                   Generated_generic_printer_base
+                   .AstPos_pat'_PConstruct_constructor constructor)
+                  #p ^^ fields_or_empty space
+            else string "(* TODO: Record? *)"
```

Comment:
This is Coq-specific, right?
Do you need any specific logic more than tuple/record/struct here?
If yes, what do you need exactly?

Reply:
This should now be reverted, and moved into Coq.